### PR TITLE
Alpha ana

### PR DESCRIPTION
**Pull request by jgalan**
jgalan Medium: 198 [](https://gitlab.cern.ch/rest-for-physics/framework/-/commits/alphaAna)

This PR is created to validate alphaAna implementation.

Related PR:
 rest-for-physics/rawlib#46
 rest-for-physics/detectorlib#33
 rest-for-physics/connectorslib#8
 rest-for-physics/tracklib#13

### CODE REVIEW

**Comment by jgalan** [comment]
PandaX-III pipeline is failing. Perhaps due to the new `TRestRawZeroSuppresionProcess`? @nkx @juanangp 

https://gitlab.cern.ch/rest-for-physics/framework/-/jobs/19174479

**Comment by juanangp** [comment]
> PandaX-III pipeline is failing. Perhaps due to the new `TRestRawZeroSuppresionProcess`? @nkx @juanangp
> 
> https://gitlab.cern.ch/rest-for-physics/framework/-/jobs/19174479

Indeed, looks like the issue is that we have to add a TRestRawToDetectorSignalProcess after TRestRawZeroSuppresionProcess.

@jgalan  are you going to fix it in panda pipeline?

**Comment by juanangp** [comment]
Another issue is that we are loosing precision because former TRestRawZeroSuppresionProcess was saving data as Double_t, while now the data is stored as a Short_t in TRestRawSignal.

**Comment by juanangp** [comment]
> Another issue is that we are loosing precision because former TRestRawZeroSuppresionProcess was saving data as Double_t, while now the data is stored as a Short_t in TRestRawSignal.

Any idea in how to fix this issue, shall we go back to the previous implementation?

**Comment by jgalan** [comment]
> Another issue is that we are loosing precision because former TRestRawZeroSuppresionProcess was saving data as Double_t, while now the data is stored as a Short_t in TRestRawSignal.

Aham, well the data in origin is in short format. So why the result is different? Perhaps at ZeroSuppression there is an additional manipulation before being transformed to `Double_t` by `TRestRawToDetectorSignalProcess`?

**Comment by jgalan** [comment]
At `TRestRawZeroSuppression` the baseline is being subtracted when invoking `GetData`.

```
        std::vector<Int_t> pOver = s->GetPointsOverThreshold();
        for (int n = 0; n < pOver.size(); n++) {
            int j = pOver[n];
            sgn.IncreaseBinBy(j, s->GetData(j));
        }      
```

We should avoid to remove the baseline as soon as we stay at `TRestRaw` domain. And we should remove it once we transform to `TRestDetectorSignal`.

So, using `GetRawData`. The problem is that when zero-ing we loss the baseline information. So we should make sure that the `TRestRawZeroSuppresionProcess` stores those values so that `TRestRawToDetectorSignalProcess` may recover them in case that specific process was used previously.

**Comment by juanangp** [comment]
> **Option 1:** We introduce those data members at `TRestRawSignal` and we move REST version to 2.4.0.
> 
> **Option 2:** We implement `zeroSuppression` as a method/option at `TRestRawToDetectorSignalProcess`.
> 
> @nkx what do you think?

I would suggest to move to option 2

**Comment by juanangp** [comment]
I really don't understand the following observabales in TRestRawZeroSuppresionProcess:

```
    SetObservableValue("flattail_map", flattailmap);
    SetObservableValue("FlatTails", totoalflatN);
    SetObservableValue("NFlatTailSignals", (int)flattailmap.size());
```
Particularly the definition:
```
int Nbefore;
        if (fNPointsFlatThreshold != 512) {
            s->InitializePointsOverThreshold(TVector2(fPointThreshold, fSignalThreshold),
                                             fNPointsOverThreshold, 512);
            Nbefore = s->GetPointsOverThreshold().size();
        }
        s->InitializePointsOverThreshold(TVector2(fPointThreshold, fSignalThreshold), fNPointsOverThreshold,
                                         fNPointsFlatThreshold);

        int Nafter = s->GetPointsOverThreshold().size();
        // cout << fRawSignalEvent->GetID() << " " << s->GetID() << " " << Nbefore << " " << Nafter << endl;
        if (Nafter != Nbefore) {
            flattailmap[s->GetID()] = Nbefore - Nafter;
            totoalflatN += Nbefore - Nafter;
        }
```
Note that `Nbefore` is not initialiazed while `fNPointsFlatThreshold==512`, which I think should be most of the cases, it doesn't make sense to me...

Shall we keep these observables or it is a good time to get rid of them?

**Comment by juanangp** [comment]
Pipeline is failing because of c++ structure bindings. So looks like pipeline is still using C++14, indeed root version is still 6.10/08 in which only C++14 was supported.

@jgalan Can we upgrade it to ROOT 6.24 and compile in  C++17?

**Comment by jgalan** [comment]
In principle I have no problem. I believe the reason behind is to keep compatibility with older systems, so installation is possible there @nkx? I guess we just need someone preparing a docker image.

We have also the LFNA GitLab instance at Unizar, where we were previously executing jobs without docker and we had much more control.

If it is not possible to keep the code compatible with both?

**Comment by juanangp** [comment]
> If it is not possible to keep the code compatible with both?

Yes, it is possible but I think we should implement C++17 features in the code, particularly structure bindings makes the coding easier.

I tested locally with C++17 and ROOT 6.24 and it is running fine to me.

**Comment by jgalan** [comment]
I am not a big fan of structures. If it is as relevant as to create a structure, perhaps a full class is a better option.

**Comment by jgalan** [comment]
I am looking at the advantages at this example https://www.geeksforgeeks.org/structured-binding-c/

And I dont really understand the benefit of linking to `x_coord`, would not be easier just to use `p.x`?

**Comment by juanangp** [comment]
Perhaps you are looking to the wrong example, structure bindings are particularly useful for e.g. std::map, for instance the part of the code that I wrote:
C++14
```
std::map<int,std::pair<int,double> > windowMap = makeMap();
  for(auto it = windowMap.begin(); it!=windowMap.end();it++){
              Double_t time = it->first * fIntWindow + fIntWindow/2.;
              Double_t energy = it->second.second/it->second.first;
            }
```
C++17
```
 std::map<int,std::pair<int,double> > windowMap = makeMap();
  for(const auto &[index, pair] : windowMap){
              Double_t time = index * fIntWindow + fIntWindow/2.;
              Double_t energy = pair.second/pair.first;
            }
```
As you can see the implementation in C++14 is not readable and error prone while in C++17 it is easy to understand.

On the other hand C++17 has many other advantages apart of structure bindings:
https://www.codingame.com/playgrounds/2205/7-features-of-c17-that-will-simplify-your-code/introduction

Still don't understand why you are against moving forward to C++17, latest version is C++20 and C++23 is on-going...

**Comment by jgalan** [comment]
> In principle I have no problem.

Again, as I said before, I am not against.

**Comment by lobis** [comment]
> In principle I have no problem. I believe the reason behind is to keep compatibility with older systems, so installation is possible on those systems @nkx? I guess we just need someone preparing a docker image.
> 
> Perhaps, we could say, v2.3.10 is compatible with C++14, v2.3.11 will not be compatible anymore?
> 
> We have also the LFNA GitLab instance at Unizar, where we were previously executing jobs without docker and we had much more control.
> 
> If it is not possible to keep the code compatible with both?

Hello,

I use this DockerHub image to do all my tests: https://hub.docker.com/r/lobis/root-geant4-garfieldpp/tags we can use this in the pipeline if you want. There are a good combination of ROOT / Geant4 / C++ versions available, I can make more if you want.

**Comment by nkx111** [comment]
> At `TRestRawZeroSuppression` the baseline is being subtracted when invoking `GetData`.
> 
> ```
>         std::vector<Int_t> pOver = s->GetPointsOverThreshold();
>         for (int n = 0; n < pOver.size(); n++) {
>             int j = pOver[n];
>             sgn.IncreaseBinBy(j, s->GetData(j));
>         }      
> ```
> 
> We should avoid to remove the baseline as soon as we stay at `TRestRaw` domain. And we should remove it once we transform to `TRestDetectorSignal`.
> 
> So, using `TRestRawSignal::GetRawData` instead of `TRestRawSignal::GetData`. The problem is that when zero-ing we loss the baseline information.
> 
> The idea behind is that we always calculate the baseline so that we do not store the values inside `TRestRawSignal`.
> 
> An option would be to include those data members `Double_t fBaseLine = 0` and `Double_t fBaseLineSigma = 0` inside `TRestRawSignal`, so that it can be recovered even after `TRestRawZeroSuppressionProcess`.
> 
> **Option 1:** We introduce those data members at `TRestRawSignal` and we move REST version to 2.4.0.
> 
> **Option 2:** We implement `zeroSuppression` as a method/option at `TRestRawToDetectorSignalProcess`.
> 
> @nkx what do you think?

I also stands for the option 2. It is too quick to move to 2.4.0 now.

**Comment by nkx111** [comment]
> I really don't understand the following observabales in TRestRawZeroSuppresionProcess:
> 
> ```
>     SetObservableValue("flattail_map", flattailmap);
>     SetObservableValue("FlatTails", totoalflatN);
>     SetObservableValue("NFlatTailSignals", (int)flattailmap.size());
> ```
> 
> Particularly the definition:
> 
> ```
> int Nbefore;
>         if (fNPointsFlatThreshold != 512) {
>             s->InitializePointsOverThreshold(TVector2(fPointThreshold, fSignalThreshold),
>                                              fNPointsOverThreshold, 512);
>             Nbefore = s->GetPointsOverThreshold().size();
>         }
>         s->InitializePointsOverThreshold(TVector2(fPointThreshold, fSignalThreshold), fNPointsOverThreshold,
>                                          fNPointsFlatThreshold);
> 
>         int Nafter = s->GetPointsOverThreshold().size();
>         // cout << fRawSignalEvent->GetID() << " " << s->GetID() << " " << Nbefore << " " << Nafter << endl;
>         if (Nafter != Nbefore) {
>             flattailmap[s->GetID()] = Nbefore - Nafter;
>             totoalflatN += Nbefore - Nafter;
>         }
> ```
> 
> Note that `Nbefore` is not initialiazed while `fNPointsFlatThreshold==512`, which I think should be most of the cases, it doesn't make sense to me...
> 
> Shall we keep these observables or it is a good time to get rid of them?

These lines are some rough handling to AGET abnormal tailing signals, developed very long time ago. I think they are already out-dated for 2.3.x. We can just remove them. In future we may develop specific processes to handle those signals. 

fNPointsFlatThreshold=512 means to turn off that functionality in `TRestRawZeroSuppresionProcess`. Here 512 is also a specialized number only for AGET electronics, which is not good.

**Comment by juanangp** [comment]
> These lines are some rough handling to AGET abnormal tailing signals, developed very long time ago. I think they are already out-dated for 2.3.x. We can just remove them. In future we may develop specific processes to handle those signals.
I have already remove them, only `nRejectedSignals` is kept while using zero suppression.

**Comment by jgalan** [comment]
You probably need to merge the present master branch into this branch.

```
git checkout master
git pull
git checkout alphaAna
git merge master
```

**Comment by juanangp** [comment]
> You probably need to merge the present master branch into this branch.
> 
> ```
> git checkout master
> git pull
> git checkout alphaAna
> git merge master
> git push
> ```

That's weird, I believe I already did in
https://github.com/rest-for-physics/framework/pull/124/commits/a4cb6a3eed63a121b2582ae2436e0ddd9e7b4527

**Comment by juanangp** [comment]
I see there is an issue now in the gitlab-ci.yml for alphaTrack. Sorry, I overlooked the pipeline failure.

I guess because we moved to a new schema? I will have a look

**Comment by lobis** [comment]
> Pipeline is failing due to C++17 structured bindings, @lobis any news on this topic?

Everything should work now with C++17 as far as I am concerned, the pipeline image uses this standard without issues.

File: source/framework/core/inc/TRestAnalysisPlot.h
Comment (lobis) [inline source/framework/core/inc/TRestAnalysisPlot.h:73]: just a general reminder that we should not be using std namespace in header files

Comment (juanangp) [reply source/framework/core/inc/TRestAnalysisPlot.h:73]: Good catch, however I cannot see `using namespace std` in the header file, I guess it is inherited from other header files.

I will create an issue about this.

**Comment by juanangp** [comment]
Finally I managed to fix the pipeline, I found something weird, in `TRestTrackReductionProcess` panda pipeline fails unless I recover `InitFromConfigFile` function https://github.com/rest-for-physics/tracklib/pull/13/commits/b86c8708565a6e66885e9579d63f90ade6551d39
Does somebody knows what is going on here? As far as I know `InitFromConfigFile` is deplecated

**Comment by jgalan** [comment]
Hi @juanangp `InitFromConfigFile` is not deprecated. But now it is implemented at `TRestMetadata` and `TRestEventProcess` directly. That means that if it is not implemented inside `TRestTrackReductionProcess` it will call to the `TRestEventProcess::InitFromConfigFile` method.

The method implemented at `TRestEventProcess::InitFromConfigFile` is capable to read basic `parameters`, but if something more complex is required, then one may re-implement `InitFromConfigFile`.

The parameter is identified with a member of the metadata class by making the first character upper case, and adding and `f` at the beginning. This is probably the reason the panda pipeline is failing. Because of the `distanceStepFactor` parameter does not correspond with the `fDistanceFactor` metadata member.

Either you modify the RML (recommended) to define `distanceFactor` or you repalce the `fDistanceStepFactor` by `fDistanceFactor` and increase the `ClassDef` version.

Present PandaX-III TrackReduction definition:
```
        <parameter name="startingDistance" value="0.15mm" />
        <parameter name="minimumDistance" value="0.5mm" />
        <parameter name="distanceStepFactor" value="1.15" />
        <parameter name="maxNodes" value="20" />
```

Present metadata members:
```
    Double_t fStartingDistance;
    Double_t fMinimumDistance;
    Double_t fDistanceFactor;
    Double_t fMaxNodes;
```

**Comment by juanangp** [comment]
When I meant deplecated I was referring that we should remove from the inherited class and use the one in the base class directly. I didn't realize that the naming doesn't match even though yesterday I was looking at it for a while...

I am going to remove  `InitFromConfigFile` and I will change the rml file. I believe I already increment the ClassDeff. I guess I don't have to increase further inside this PR.

**Comment by jgalan** [comment]
> I didn't realize that the naming doesn't match even though yesterday I was looking at it for a while...

I know the feeling, it happened to me many times. I guess the problem will not rise again once `InitFromConfigFile` is removed from most classes.

Anyhow those know-how issues should be soon documented at the user and developer guide introduction to REST.

https://rest-for-physics.github.io/rest-advanced/metadata-development.html

**Review by juanangp — APPROVED** [review]

**Comment by juanangp** [comment]
@lobis and @jgalan this PR is hanging since a while, can you please have a look to it, I would like to merge soon.

Also check the related PR:
https://github.com/rest-for-physics/rawlib/pull/46
https://github.com/rest-for-physics/detectorlib/pull/33
https://github.com/rest-for-physics/connectorslib/pull/8
https://github.com/rest-for-physics/tracklib/pull/13

**Comment by jgalan** [comment]
Yes, but I cannot approve it myself since I created. It seems fine to me to merge it.

**Comment by jgalan** [comment]
https://github.com/rest-for-physics/rawlib/pull/46 pipeline is failing

**Comment by juanangp** [comment]
> [rest-for-physics/rawlib#46](https://github.com/rest-for-physics/rawlib/pull/46) pipeline is failing

This is because the docker image is not the right one, I think should be fine after merging. Similar issues have been spotted in other PR.

**Comment by juanangp** [comment]
> > [rest-for-physics/rawlib#46](https://github.com/rest-for-physics/rawlib/pull/46) pipeline is failing
> 
> This is because the docker image is not the right one, I think should be fine after merging. Similar issues have been spotted in other PR.

In fact this is fixed in  [rest-for-physics/rawlib#51](https://github.com/rest-for-physics/rawlib/pull/51) which is not yet merged in master.

**Comment by jgalan** [comment]
> > > [rest-for-physics/rawlib#46](https://github.com/rest-for-physics/rawlib/pull/46) pipeline is failing
> > 
> > 
> > This is because the docker image is not the right one, I think should be fine after merging. Similar issues have been spotted in other PR.
> 
> In fact this is fixed in [rest-for-physics/rawlib#51](https://github.com/rest-for-physics/rawlib/pull/51) which is not yet merged in master.

But rest-for-physics/rawlib#51 is also failing

**Comment by jgalan** [comment]
Actually rawlib at master is also failing after merging rest-for-physics/rawlib#50. Perhaps we just need to update the docker image? @lobis

**Comment by juanangp** [comment]
> > > > [rest-for-physics/rawlib#46](https://github.com/rest-for-physics/rawlib/pull/46) pipeline is failing
> > > 
> > > 
> > > This is because the docker image is not the right one, I think should be fine after merging. Similar issues have been spotted in other PR.
> > 
> > 
> > In fact this is fixed in [rest-for-physics/rawlib#51](https://github.com/rest-for-physics/rawlib/pull/51) which is not yet merged in master.
> 
> But [rest-for-physics/rawlib#51](https://github.com/rest-for-physics/rawlib/pull/51) is also failing

Indeed, this is another un-related issue to this PR after the migration to C++17. This issue arises from performing on-the-fly compilation of STL containers and hardcoding compilation flags in the code:
https://github.com/rest-for-physics/framework/blob/1d60c1048bbccd839ee2e8e783bb5309d0bd9abc/source/framework/tools/src/TRestReflector.cxx#L230 
A patch could be
```
system(Form("gcc %s `root-config --cflags` "
```
@lobis we had a chat about this, we don't have any related issue?

**Comment by jgalan** [comment]
However, the rawlib pipeline seems to fail at the `cmake` level.

**Comment by juanangp** [comment]
> However, the rawlib pipeline seems to fail at the `cmake` level.

master is failing because we don't have the proper docker image.

Docker image is updated on https://github.com/rest-for-physics/rawlib/pull/51, however is failing becasue of https://github.com/rest-for-physics/framework/blob/1d60c1048bbccd839ee2e8e783bb5309d0bd9abc/source/framework/tools/src/TRestReflector.cxx#L230

**Comment by jgalan** [comment]
So, then why dont try to update it at that PR using **system(Form("gcc %s `root-config --cflags` "** and see if it goes through? I guess it makes sense, so that we do not force c++11.

**Review by lobis — APPROVED** [review]

File: cmake/FindGarfieldOld.cmake
Comment (juanangp) [reply cmake/FindGarfieldOld.cmake]: I believe this line is not needed and it was actually removed, I don't understand how it end up in this PR.

As fas as I remember some commits were cherry-pick from this PR to master. Perhaps something when wrong in between.

Comment (jgalan) [reply cmake/FindGarfieldOld.cmake]: This branch is behind master, perhaps it will be fixed once master is merged in this branch. Highly recommended.

Screenshot 2022-04-20 at 21 35 51

Comment (jgalan) [reply cmake/FindGarfieldOld.cmake]: I guess thats why "outdated" appears next to the file name

**Review by juanangp — APPROVED** [review]

**Review by juanangp — APPROVED** [review]

**Review by lobis — APPROVED** [review]

**Pull request merged**
